app/features/build_runner/remote_runner.rb

```diff
@@ -20,187 +28,193 @@ class RemoteRunner
-        fsm.when(:RUNNING,   PENDING:   :RUNNING)
-        fsm.when(:REJECTED,  PENDING:   :REJECTED)
-        fsm.when(:FINISHING, RUNNING:   :FINISHING)
-        fsm.when(:SUCCEEDED, FINISHING: :SUCCEEDED)
-        fsm.when(:FAILED,    RUNNING:   :FAILED)
-        fsm.when(:BROKEN,    PENDING:   :BROKEN,
-                             RUNNING:   :BROKEN,
-                             FINISHING: :BROKEN)
-      end
-    end
-
-    # Add a listener to get real time updates on new rows (see `new_row`)
-    # This is used for the socket connection to the user's browser
-    def add_build_change_listener(block)
-      @build_change_observer_blocks << block
+      @history = [] # TODO: consider implications if history becomes really large.
+      @complete = false
+      @completion_blocks = [] # TODO(snatchev): does this needs to be a threadsafe array?
     end
 
     def start
-      env = environment_variables_for_worker
+      save_build_status_locally!
 
+      env = environment_variables_for_worker
       start_time = Time.now.utc
+
       begin
-        responses = Agent::Client.new("localhost").request_run_fastlane(
-          "bundle", "exec", "fastlane", project.platform, project.lane, env: env
-        )
+        responses = @grpc.request_run_fastlane("bundle", "exec", "fastlane", project.platform, project.lane, env: env)
+        return if responses.nil?
+
         responses.each do |response|
-          validate_agent_response(response: response)
-          process_agent_response(response: response)
+          # update the build's duration every time we get a message from the runner.
+          current_build.duration = Time.now.utc - start_time
+
+          # dispatch to handler methods
+          handle_log(response.log)           if response.log
+          handle_state(response.state)       if response.state
+          handle_error(response.error)       if response.error
+          handle_artifact(response.artifact) if response.artifact
         end
-      rescue StandardError => exception
-        emit_error_response(
-          FastlaneCI::Proto::InvocationResponse::Error.new(
-            description: exception.message,
-            stacktrace: exception.backtrace.join("\n")
-          )
-        )
-        emit_new_row(
-          type: :ERROR,
-          message: "Please make sure that the build agent is running!",
-          time: Time.now
-        )
+      rescue GRPC::Unavailable, GRPC::DeadlineExceeded
+        logger.error("Agent is not running or did not respond on #{@grpc.host}:#{@grpc.port}")
+        handle_error(FastlaneCI::Proto::InvocationResponse::Error.new(description: "The Agent is not available"))
+      ensure
+        @complete = true
+        @completion_blocks.each(&:call)
       end
-      emit_new_row(type: :last_message, message: nil, time: Time.now)
 
-      current_build.duration = Time.now.utc - start_time
-      if @build_state.state == :SUCCEEDED
+      persist_history!
```

Comment:
> This doesn't happen if the agent dies - and I believe we loose the logs we got that far in that case.

We can move it to the `ensure` block, and I believe that will get called if the agent dies during the `begin/rescue/ensure` block.

> Side note - I feel like we should save log lines to runner.log everytime we get one, rather than at the end.

That's how it was originally, but then noticed that in those cases, we would only persist logs, and would not get state or errors.